cmd/uniq/uniq.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+)
+
+type options struct {
+	printDuplicates bool
+	printEmptyLines bool
+	printLineNumber bool
+}
+
+func parseArgs(args []string) options {
+	var opts options
+
+	for _, opt := range args[1:] {
+		switch opt {
+		case "-dup":
+			opts.printDuplicates = true
+		case "-empty":
+			opts.printEmptyLines = true
+		case "-num":
+			opts.printLineNumber = true
+		default:
+			fmt.Println("Usage:")
+			fmt.Println("uniq [-dup | -empty | -num]")
+			os.Exit(1)
+		}
+	}
+	return opts
+}
+
+func uniq(input *os.File, opts options) {
```

Comment:
why everything in the same function? Too much context, lots of variables in same scope, etc. Hard to read.

Breaking in several functions do not hurt, functions are inlined in some cases. Maps and slices are references, never copied if passed as parameters or returned from functions.